hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/BlockPlacementPolicyRackFaultTolerant.java

```diff
@@ -192,11 +192,19 @@ private void chooseEvenlyFromRemainingRacks(Node writer,
       } finally {
         excludedNodes.addAll(newExcludeNodes);
       }
+      if (numResultsOflastChoose == results.size()) {
+        Map<String, Integer> nodesPerRack = new HashMap<>();
+        for (DatanodeStorageInfo dsInfo : results) {
+          String rackName = dsInfo.getDatanodeDescriptor().getNetworkLocation();
+          nodesPerRack.merge(rackName, 1, Integer::sum);
+        }
+        bestEffortMaxNodesPerRack = Collections.max(nodesPerRack.values());
```

Comment:
Is it possible to introduce infinite loops here? If each rack already has one chosen node and `bestEffortMaxNodesPerRack` is 2, and no datanode can be chosen now, then `bestEffortMaxNodesPerRack`  will change to 1 after line 201, which may cause an infinite loop. So the calculation of the maximum value should consider the old value of `bestEffortMaxNodesPerRack` to insure increase.

Reply:
Thank you for your feedback. I will make the necessary code changes as soon as possible.